src/Functions/FunctionGenerateRandomStructure.cpp

```diff
@@ -84,119 +93,147 @@ class FunctionGenerateRandomStructure : public IFunction
+    size_t generateNumberOfColumns(pcg64 & rng)
+    {
+        return rng() % MAX_NUMBER_OF_COLUMNS + 1;
     }
 
-    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    void writeLowCardinalityNestedType(pcg64 & rng, WriteBuffer & buf, bool allow_suspicious_lc_types, bool allow_nullable = true)
     {
-        size_t seed = randomSeed();
-        size_t number_of_columns = 0;
+        bool make_nullable = allow_nullable & rng() % 2;
+        if (make_nullable)
+            writeCString("Nullable(", buf);
 
-        if (!arguments.empty() && !arguments[0].column->onlyNull())
+        if (allow_suspicious_lc_types)
         {
-            number_of_columns = arguments[0].column->getUInt(0);
-            if (number_of_columns > MAX_NUMBER_OF_COLUMNS)
-                throw Exception(
-                    ErrorCodes::BAD_ARGUMENTS,
-                    "Maximum allowed number of columns is {}, got {}",
-                    MAX_NUMBER_OF_COLUMNS,
-                    number_of_columns);
+            TypeIndex type = suspicious_lc_types[rng() % map_key_types.size()];
+            if (type == TypeIndex::FixedString)
+                writeString("FixedString(" + std::to_string(rng() % MAX_FIXEDSTRING_SIZE_WITHOUT_SUSPICIOUS + 1) + ")", buf);
+            else
+                writeString(magic_enum::enum_name<TypeIndex>(type), buf);
+        }
+        else
+        {
+            /// Support only String and FixedString.
+            if (rng() % 2)
+                writeCString("String", buf);
+            else
+                writeString("FixedString(" + std::to_string(rng() % MAX_FIXEDSTRING_SIZE_WITHOUT_SUSPICIOUS + 1) + ")", buf);
         }
 
-        if (arguments.size() > 1 && !arguments[1].column->onlyNull())
-            seed = arguments[1].column->getUInt(0);
+        if (make_nullable)
+            writeChar(')', buf);
+    }
 
-        pcg64 rng(seed);
-        if (number_of_columns == 0)
-            number_of_columns = generateNumberOfColumns(rng);
 
-        auto col_res = ColumnString::create();
-        auto & string_column = assert_cast<ColumnString &>(*col_res);
-        auto & chars = string_column.getChars();
-        WriteBufferFromVector buf(chars);
-        for (size_t i = 0; i != number_of_columns; ++i)
+    void writeEnumValues(const String & column_name, pcg64 & rng, WriteBuffer & buf, ssize_t max_value)
+    {
+        /// Don't generate big enums, because it will lead to really big result
+        /// and slowness of this function, and it can lead to `Max query size exceeded`
+        /// while using this function with generateRandom.
+        size_t num_values = rng() % 16 + 1;
+        std::vector<Int16> values(num_values);
+
+        /// Generate random numbers from range [-(max_value + 1), max_value - num_values + 1]
+        for (Int16 & x : values)
+            x = rng() % (2 * (max_value + 1) - num_values) - max_value - 1;
```

Comment:
Off by one, I think. If num_values = 1, range of x should be [-(max_value - 1), max_value].
```suggestion
            x = rng() % (2 * max_value + 3 - num_values) - max_value - 1;
```